src/main/scala/io/iohk/ethereum/domain/BlockBody.scala

```diff
@@ -3,19 +3,136 @@ package io.iohk.ethereum.domain
 import io.iohk.ethereum.domain.BlockHeaderImplicits._
 import io.iohk.ethereum.rlp.{RLPEncodeable, RLPList, RLPSerializable, rawDecode}
 
-case class BlockBody(transactionList: Seq[SignedTransaction], uncleNodesList: Seq[BlockHeader]) {
-  override def toString: String =
-    s"BlockBody{ transactionList: $transactionList, uncleNodesList: $uncleNodesList }"
+import scala.collection.immutable
+
+class BlockBody private(
+                        private val transactionList: Array[SignedTransaction],
+                        val uncleNodesList: Seq[BlockHeader]
+                      ) extends Iterable[SignedTransaction] {
+
 
-  lazy val numberOfTxs: Int = transactionList.size
+  lazy val numberOfTxs: Int = transactionList.length
 
   lazy val numberOfUncles: Int = uncleNodesList.size
+
+  /**
+    * Works as `copy` for case classes
+    * @param transactions - new transactions for the block
+    *        Beware using array-based collections as an argument, because:
+    *  {{{
+    *  val ar1 = Array(1, 2, 3)
+    *  val seq = ar1.toSeq
+    *  val ar2 = seq.toArray
+    *  ar2.update(0,0)
+    *  ar1
+    *  }}}
+    *  would return (0, 2, 3), because ar1 === ar2
+    * @return new BlockBody with same uncles but new `transactions`.
+    */
+  def withTransactions(transactions: Seq[SignedTransaction]): BlockBody = {
+    new BlockBody(transactions.toArray, this.uncleNodesList)
+  }
+
+  /**
+    * @param uncles - new uncles for the block
+    * @return new BlockBody with same `transactions` but replaced uncles.
+    */
+  def withUncles(uncles: Seq[BlockHeader]): BlockBody = {
+    BlockBody(this.transactionList, uncles)
+  }
+
+  /**
+   * @return An inner array size.
+   */
+  override def size: Int = numberOfTxs
+
+
+  override def iterator: Iterator[SignedTransaction] = {
```

Comment:
i would rename it to `transactionIteratator` (and do that with other methods which operates on signed transactions) as blockbody has also `unclesList` so it is not obvious which iterator we are getting.